### PR TITLE
interface: Taints & Tolerations API

### PR DESCRIPTION
**Pull request by Arvindthiru**
### Description of your changes

Fixes #

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### Special notes for your reviewer

Decided to go with custom objects for taints and tolerations because,

- The objects from "k8s.io/api/core/v1" were specifically created for nodes, pods and hence has comments explicitly specifying they were related to nodes and pods 
image
image
    - Taint had TimeAdded field for NoExecute 
image
    - Toleration has TolerationSeconds field for NoExecute 
image

### CODE REVIEW

File: apis/cluster/v1beta1/membercluster_types.go [hunk 1]
```diff
@@ -69,6 +74,21 @@ type MemberClusterStatus struct {
...
+	// The taint key to be applied to a MemberCluster.
+	// +required
+	Key string `json:"key"`
+	// The taint value corresponding to the taint key.
+	// +optional
+	Value string `json:"value,omitempty"`
+	// The effect of the taint on ClusterResourcePlacements that do not tolerate the taint.
+	// Valid effect is NoSchedule.
+	// +required
+	Effect corev1.TaintEffect `json:"effect"`
```
Comment (zhiying-lin) [inline apis/cluster/v1beta1/membercluster_types.go:89]: can you add kubebuilder validation marker?

eg, 
//+kubebuilder:validation:Enum=PickAll;PickN;PickFixed

File: apis/cluster/v1beta1/membercluster_types.go [hunk 2]
```diff
@@ -69,6 +74,21 @@ type MemberClusterStatus struct {
...
+// any ClusterResourcePlacement that does not tolerate the Taint.
+type Taint struct {
+	// The taint key to be applied to a MemberCluster.
+	// +required
+	Key string `json:"key"`
+	// The taint value corresponding to the taint key.
+	// +optional
+	Value string `json:"value,omitempty"`
+	// The effect of the taint on ClusterResourcePlacements that do not tolerate the taint.
+	// Valid effect is NoSchedule.
```
Comment (zhiying-lin) [inline apis/cluster/v1beta1/membercluster_types.go:87]:
```suggestion
	// Only NoSchedule is supported.
```

File: apis/placement/v1beta1/clusterresourceplacement_types.go [hunk 3]
```diff
@@ -168,6 +169,10 @@ type PlacementPolicy struct {
 	// +patchMergeKey=topologyKey
 	// +patchStrategy=merge
 	TopologySpreadConstraints []TopologySpreadConstraint `json:"topologySpreadConstraints,omitempty" patchStrategy:"merge" patchMergeKey:"topologyKey"`
+
+	// If specified, the ClusterResourcePlacement's tolerations.
+	// +optional
+	Tolerations []Toleration `json:"tolerations,omitempty"`
```
Comment (zhiying-lin) [inline apis/placement/v1beta1/clusterresourceplacement_types.go:175]: Set the maxItems?

// +kubebuilder:validation:MaxItems=xx

File: apis/placement/v1beta1/clusterresourceplacement_types.go [hunk 4]
```diff
@@ -455,6 +460,29 @@ type FailedResourcePlacement struct {
...
+	// +optional
+	Operator corev1.TolerationOperator `json:"operator,omitempty"`
+	// Value is the taint value the toleration matches to.
+	// If the operator is Exists, the value should be empty, otherwise just a regular string.
+	// +optional
+	Value string `json:"value,omitempty"`
+	// Effect indicates the taint effect to match. Empty means match all taint effects.
+	// When specified, allowed value is NoSchedule.
+	// +optional
+	Effect corev1.TaintEffect `json:"effect,omitempty"`
```
Comment (zhiying-lin) [inline apis/placement/v1beta1/clusterresourceplacement_types.go:483]: Add the kubebuilder validation marker?

File: apis/placement/v1beta1/clusterresourceplacement_types.go [hunk 5]
```diff
@@ -455,6 +460,29 @@ type FailedResourcePlacement struct {
...
+	// Key is the taint key that the toleration applies to. Empty means match all taint keys.
+	// If the key is empty, operator must be Exists; this combination means to match all values and all keys.
+	// +optional
+	Key string `json:"key,omitempty"`
+	// Operator represents a key's relationship to the value.
+	// Valid operators are Exists and Equal. Defaults to Equal.
+	// Exists is equivalent to wildcard for value, so that a pod can
+	// tolerate all taints of a particular category.
+	// +optional
+	Operator corev1.TolerationOperator `json:"operator,omitempty"`
```
Comment (zhiying-lin) [inline apis/placement/v1beta1/clusterresourceplacement_types.go:475]: add the kubebuilder validation and default marker?

File: apis/cluster/v1beta1/membercluster_types.go [hunk 6]
```diff
@@ -50,6 +50,7 @@ type MemberClusterSpec struct {
 	HeartbeatPeriodSeconds int32 `json:"heartbeatPeriodSeconds,omitempty"`
 
 	// If specified, the MemberCluster's taints.
```
Comment (zhiying-lin) [inline apis/cluster/v1beta1/membercluster_types.go:52]: do you want to add a comment mention that we don't allow update/delete taints and only adding is allowed?

**Review by zhiying-lin — COMMENTED** [review]
one minor comment, others LGTM

File: apis/cluster/v1beta1/membercluster_types.go (hunk 6 above)
Comment (ryanzhang-oss) [reply apis/cluster/v1beta1/membercluster_types.go:52]: no delete after a taint is set on the cluster? I don't recall k8s has this behavior.

File: apis/placement/v1beta1/clusterresourceplacement_types.go [hunk 7]
```diff
@@ -455,6 +462,32 @@ type FailedResourcePlacement struct {
...
 }
 
+// Toleration allows ClusterResourcePlacement to tolerate any taint that matches
+// the triple <key,value,effect> using the matching operator <operator>.
+type Toleration struct {
+	// Key is the taint key that the toleration applies to. Empty means match all taint keys.
+	// If the key is empty, operator must be Exists; this combination means to match all values and all keys.
+	// +optional
+	Key string `json:"key,omitempty"`
+	// Operator represents a key's relationship to the value.
```
Comment (ryanzhang-oss) [inline apis/placement/v1beta1/clusterresourceplacement_types.go:472]: nit: space?

File: apis/cluster/v1beta1/membercluster_types.go (hunk 6 above)
Comment (Arvindthiru) [reply apis/cluster/v1beta1/membercluster_types.go:52]: I synced with Zhiying offline this comment was for tolerations and I've addresed it

**Pull request merged**